File: lms/www/billing/billing.js
frappe.ready(() => {
	if ($("#billing-form").length) {
		frappe.require("controls.bundle.js", () => {
			setup_billing();
		});
	}

	$(".btn-pay").click((e) => {
		generate_payment_link(e);
	});
});

const setup_billing = () => {
	this.billing = new frappe.ui.FieldGroup({
		fields: [
			{
				fieldtype: "Data",
				label: __("Billing Name"),
				fieldname: "billing_name",
				reqd: 1,
				default: address && address.billing_name,
			},
			{
				fieldtype: "Data",
				label: __("Address Line 1"),
				fieldname: "address_line1",
				reqd: 1,
				default: address && address.address_line1,
			},
			{
				fieldtype: "Data",
				label: __("Address Line 2"),
				fieldname: "address_line2",
				default: address && address.address_line2,
			},
			{
				fieldtype: "Data",
				label: __("City/Town"),
				fieldname: "city",
				reqd: 1,
				default: address && address.city,
			},
			{
				fieldtype: "Data",
				label: __("State/Province"),
				fieldname: "state",
				default: address && address.state,
			},
			{
				fieldtype: "Column Break",
			},
			{
				fieldtype: "Link",
				label: __("Country"),
				fieldname: "country",
				options: "Country",
				reqd: 1,
				only_select: 1,
				default: address && address.country,
				change: () => {
					change_currency();
				},
			},
			{
				fieldtype: "Data",
				label: __("Postal Code"),
				fieldname: "pincode",
				reqd: 1,
				default: address && address.pincode,
			},
			{
				fieldtype: "Data",
				label: __("Phone Number"),
				fieldname: "phone",
				reqd: 1,
				default: address && address.phone,
			},
			{
				fieldtype: "Link",
				label: __("Where did you hear about this?"),
				fieldname: "source",
				options: "LMS Source",
				only_select: 1,
				reqd: 1,
			},
			{
				fieldtype: "Section Break",
				label: __("GST Details"),
				fieldname: "gst_details",
				depends_on: "eval:doc.country === 'India'",
			},
			{
				fieldtype: "Data",
				label: __("GSTIN"),
				fieldname: "gstin",
			},
			{
				fieldtype: "Column Break",
				fieldname: "gst_details_break",
			},
			{
				fieldtype: "Data",
				fieldname: "pan",
				label: __("PAN"),
			},
		],
		body: $("#billing-form").get(0),
	});
	this.billing.make();
	$("#billing-form .form-section:last").removeClass("empty-section");
	$("#billing-form .frappe-control").removeClass("hide-control");
	$("#billing-form .form-column").addClass("p-0");
};

const generate_payment_link = (e) => {
	let new_address = this.billing.get_values();
	validate_address(new_address);
	let doctype = $(e.currentTarget).attr("data-doctype");
	let docname = decodeURIComponent($(e.currentTarget).attr("data-name"));

	frappe.call({
		method: "lms.lms.utils.get_payment_options",
		args: {
			doctype: doctype,
			docname: docname,
			phone: new_address.phone,
			country: new_address.country,
		},
		callback: (data) => {
			data.message.handler = (response) => {
				handle_success(
					response,
					doctype,
					docname,
					new_address,
					data.message.order_id
				);
			};
			let rzp1 = new Razorpay(data.message);
			rzp1.open();
		},
	});
};

const handle_success = (response, doctype, docname, address, order_id) => {
	frappe.call({
		method: "lms.lms.utils.verify_payment",
		args: {
			response: response,
			doctype: doctype,
			docname: docname,
			address: address,
			order_id: order_id,
		},
		callback: (data) => {
			frappe.show_alert({
				message: __("Payment Successful"),
				indicator: "green",
			});
			setTimeout(() => {
				window.location.href = data.message;
			}, 1000);
		},
	});
};

const change_currency = () => {
	$("#gst-message").removeClass("hide");
	let country = this.billing.get_value("country");
	if (exception_country.includes(country)) {
		update_price(original_price_formatted);
		return;
	}
	frappe.call({
		method: "lms.lms.utils.change_currency",
		args: {
			country: country,
			amount: amount,
			currency: currency,
		},
		callback: (data) => {
			let current_price = $(".total-price").text();
			if (current_price != data.message) {
				update_price(data.message);
			}
			if (data.message.includes("INR")) {
				$("#gst-message").removeClass("hide").addClass("show");
			} else {
				$("#gst-message").removeClass("show").addClass("hide");
			}
		},
	});
};

const update_price = (price) => {
	$(".total-price").text(price);
	frappe.show_alert({
		message: "Total Price has been updated.",
		indicator: "yellow",
	});
};

const validate_address = (billing_address) => {
	if (billing_address.country == "India" && !billing_address.state)
		frappe.throw(__("State is mandatory."));

	const states = [
		"Andhra Pradesh",
		"Arunachal Pradesh",
		"Assam",
		"Bihar",
		"Chhattisgarh",
		"Goa",
		"Gujarat",
		"Haryana",
		"Himachal Pradesh",
		"Jharkhand",
		"Karnataka",
		"Kerala",
		"Madhya Pradesh",
		"Maharashtra",
		"Manipur",
		"Meghalaya",
		"Mizoram",
		"Nagaland",
		"Odisha",
		"Punjab",
		"Rajasthan",
		"Sikkim",
		"Tamil Nadu",
		"Telangana",
		"Tripura",
		"Uttar Pradesh",
		"Uttarakhand",
		"West Bengal",
	];
	if (
		billing_address.country == "India" &&
		!states.includes(billing_address.state)
	)
		frappe.throw(
			__(
				"Please enter a valid state with correct spelling and the first letter capitalized."
			)
		);
};

File: lms/www/billing/billing.py
import frappe
from frappe import _
from lms.lms.utils import check_multicurrency, apply_gst


def get_context(context):
	module = frappe.form_dict.module
	docname = frappe.form_dict.modulename
	doctype = "LMS Course" if module == "course" else "LMS Batch"

	context.module = module
	context.docname = docname
	context.doctype = doctype

	validate_access(doctype, docname, module)
	get_billing_details(context)

	context.original_currency = context.currency
	context.original_amount = (
		apply_gst(context.amount, None)[0]
		if context.original_currency == "INR"
		else context.amount
	)

	context.exception_country = frappe.get_all(
		"Payment Country", filters={"parent": "LMS Settings"}, pluck="country"
	)

	context.amount, context.currency = check_multicurrency(
		context.amount, context.currency
	)

	context.address = get_address()
	if context.currency == "INR":
		context.amount, context.gst_applied = apply_gst(context.amount, None)


def validate_access(doctype, docname, module):
	if frappe.session.user == "Guest":
		raise frappe.PermissionError(_("Please login to continue with payment."))

	if module not in ["course", "batch"]:
		raise ValueError(_("Module is incorrect."))

	if not frappe.db.exists(doctype, docname):
		raise ValueError(_("Module Name is incorrect or does not exist."))

	if doctype == "LMS Course":
		membership = frappe.db.exists(
			"LMS Enrollment", {"member": frappe.session.user, "course": docname}
		)
		if membership:
			raise frappe.PermissionError(_("You are already enrolled for this course"))

	else:
		membership = frappe.db.exists(
			"Batch Student", {"student": frappe.session.user, "parent": docname}
		)
		if membership:
			raise frappe.PermissionError(_("You are already enrolled for this batch."))


def get_billing_details(context):
	if context.doctype == "LMS Course":
		details = frappe.db.get_value(
			"LMS Course",
			context.docname,
			["title", "name", "paid_course", "course_price as amount", "currency"],
			as_dict=True,
		)

		if not details.paid_course:
			raise frappe.PermissionError(_("This course is free."))

	else:
		details = frappe.db.get_value(
			"LMS Batch",
			context.docname,
			["title", "name", "paid_batch", "amount", "currency"],
			as_dict=True,
		)

		if not details.paid_batch:
			raise frappe.PermissionError(
				_("To join this batch, please contact the Administrator.")
			)

	context.title = details.title
	context.amount = details.amount
	context.currency = details.currency


def get_address():
	address = frappe.get_all(
		"Address",
		{"email_id": frappe.session.user},
		[
			"address_title as billing_name",
			"address_line1",
			"address_line2",
			"city",
			"state",
			"country",
			"pincode",
			"phone",
		],
		order_by="creation desc",
		limit=1,
	)

	if not len(address):
		return None
	else:
		address = address[0]

	if not address.address_line2:
		address.address_line2 = ""

	if not address.state:
		address.state = ""

	return address
